nengo_loihi/loihi_api.py

```diff
@@ -19,10 +19,56 @@
 BIAS_EXP_MAX = 2**3 - 1
 BIAS_MAX = BIAS_MAN_MAX * 2**BIAS_EXP_MAX
 
-U_MAX = 2**23 - 1
-U_MIN = -2**23
-V_MAX = 2**23 - 1
-V_MIN = -2**23
+Q_BIT = 21  # number of bits for synapse accumulator
+U_BIT = 23  # number of bits for cx input (u)
+# V_BIT = 24  # number of bits for cx voltage (v)
+
+
+def overflow_signed(x, bits=7, out=None, return_hits=False):
```

Comment:
I think we need some unit tests for this function.  It seems like there are a lot of different conditions to worry about, and a lot of bit math with potential to go awry.  `test_uv_overflow` checks that the emulator matches the hardware given certain inputs, but it's hard to know whether that is covering all the potential overflow cases.  In addition, it would be nice to assert concrete behaviour (e.g. given these inputs, this is what the output should be), rather than just matching to the hardware.  That way we know that the overflow behaviour is working exactly the way we think it is, as opposed to having to wonder whether some mismatch is due to a problem in our emulator, a problem in the hardware, or a problem in this function.

Reply:
Any ideas on how to write a unit test for this? I'm not quite sure how to write a general test for it, without just replicating the code in the function. I guess I could just write it without all the bit twiddling optimizations that I put in.

Reply:
You just want to test specific cases that cover the different overflow possibilities.  Something like 
```
# basic check
assert(overflow_signed(16, bits=2) == ?
# check that negative values are handled appropriately
assert overflow_signed(-16, bits=2 == ? 
# check that changing the value of bits changes the overflow as we expect
assert overflow_signed(16, bits=3) == ?
...
```
and so on to check other possibilities (e.g. other values for x? is zero handled correctly? is `return_hits` correct in all these cases?).  You probably know better than me the different ways in which these values could overflow, and what edge cases we'd need to worry about.

Reply:
Ok, I can do something like that. In my experience, though, it's hard to predict the edge cases without having a robust programmatic test that tests it for thousands/millions of potential values (you never know when you'll hit something like the 300 bug). But I'll put in what I can think of.